frame/staking/src/ledger.rs

```diff
@@ -0,0 +1,375 @@
+/// the staking locks accordingly.
+/// - `StakingLedger::storage_exists`: checks if an account has staking ledger entry in [`Ledger`].
+#[derive(
+	PartialEqNoBound,
+	EqNoBound,
+	CloneNoBound,
+	Encode,
+	Decode,
+	RuntimeDebugNoBound,
+	TypeInfo,
+	MaxEncodedLen,
+)]
+#[scale_info(skip_type_params(T))]
+pub struct StakingLedger<T: Config> {
+	/// The stash account whose balance is actually locked and at stake.
+	pub stash: T::AccountId,
+	/// The total amount of the stash's balance that we are currently accounting for.
+	/// It's just `active` plus all the `unlocking` balances.
+	#[codec(compact)]
+	pub total: BalanceOf<T>,
+	/// The total amount of the stash's balance that will be at stake in any forthcoming
+	/// rounds.
+	#[codec(compact)]
+	pub active: BalanceOf<T>,
+	/// Any balance that is becoming free, which may eventually be transferred out of the stash
+	/// (assuming it doesn't get slashed first). It is assumed that this will be treated as a first
+	/// in, first out queue where the new (higher value) eras get pushed on the back.
+	pub unlocking: BoundedVec<UnlockChunk<BalanceOf<T>>, T::MaxUnlockingChunks>,
+	/// List of eras for which the stakers behind a validator have claimed rewards. Only updated
+	/// for validators.
+	pub claimed_rewards: BoundedVec<EraIndex, T::HistoryDepth>,
+	/// The controller associated with this ledger's stash.
+	///
+	/// This is not stored on-chain, and is only bundled when the ledger is read from storage.
+	/// Use [`controller`] function to get the controller associated with the ledger.
+	#[codec(skip)]
+	controller: Option<T::AccountId>,
+}
+
+impl<T: Config> StakingLedger<T> {
+	#[cfg(feature = "runtime-benchmarks")]
+	pub fn default_from(stash: T::AccountId) -> Self {
+		Self {
+			stash,
+			total: Zero::zero(),
+			active: Zero::zero(),
+			unlocking: Default::default(),
+			claimed_rewards: Default::default(),
+			controller: Default::default(),
+		}
+	}
+
+	/// Returns a new instance of a staking ledger.
+	///
+	/// The `Ledger` storage is not mutated. In order to do that [`fn storage_put`] must be called
+	/// on the returned staking ledger.
+	///
+	/// Note: as the controller accounts are being deprecated, the stash account is the same as the
+	/// controller account.
+	pub fn new(
+		stash: T::AccountId,
+		active_stake: BalanceOf<T>,
+		total_stake: BalanceOf<T>,
+		unlocking: BoundedVec<UnlockChunk<BalanceOf<T>>, T::MaxUnlockingChunks>,
+		claimed_rewards: BoundedVec<EraIndex, T::HistoryDepth>,
+	) -> Self {
+		Self {
+			stash: stash.clone(),
+			active: active_stake,
+			total: total_stake,
+			unlocking,
+			claimed_rewards,
+			// controllers are deprecated and map 1-1 to stashes.
+			controller: Some(stash),
+		}
+	}
+
+	/// Returns the controller account of the staking ledger.
+	fn controller(&self) -> T::AccountId {
+		self.controller
+			.clone()
+			.expect("staking ledger instance was fetched through `storage_get`, qed.")
```

Comment:
I'm not 100% happy with this expect, since it will panic if `self` was fetched directly through `Ledger<T>` without using `fn storage_get`. In the future this may be a source of issues and silent runtime inconsistencies. 

I may refactor this to return an option, but that will add a bit more complexity to the caller's logic.

Reply:
I agree, done!